### PR TITLE
ci: Use uv pip for faster build

### PR DESCRIPTION
**Pull request by rht**
[`uv`](https://astral.sh/blog/uv) is a fast drop-in `pip`, `pip-compile`, `virtualenv` etc replacement created by the creator of Ruff.

### CODE REVIEW

**Comment by Corvince** [comment]
Thank you for mentioning uv, the news haven't reached me yet. I think this is quite exciting and I will start using uv today, but I strongly discourage its use for mesa just yet. The first public release is less than 24 hours ago. This potentially decreases our workflows by a few seconds, but I don't think that justifies the possible hurdles that are attached to bleeding edge software.

**Comment by rht** [comment]
I have yet to measure by how much this shaves off the build time, because the CI still fails.

According to the blog post:
> [uv](https://github.com/astral-sh/uv) is designed as a drop-in replacement for pip and pip-tools, and is ready for production use today in projects built around those workflows.

If it builds and runs the code just fine, I see no reason not to use it. Besides, Ruff is considered bleeding edge, yet many huge projects use it in their CI. Not all "bleeding edge" public-released tools are the same in terms of reliability.

**Comment by rht** [comment]
I have overcome https://github.com/astral-sh/uv/issues/1374. Waiting for https://github.com/astral-sh/uv/issues/1355 to be fixed, which should happen really soon.

**Comment by Corvince** [comment]
>  Not all "bleeding edge" public-released tools are the same in terms of reliability.

Yes but there have already been 2 bugfix releases in the first 12 hours of its existence and you already linked two issues that you had to overcome/waiting for a fix. And these are just the obvious bugs, I am more worried about the subtle bugs, where it appears to be working, but some details are broken. This is inevitable for such large projects. We don't know if it will affect us, but why risk it?

**Comment by rht** [comment]
Didn't you want to know how much this will speed up the macOS and Windows build? My sense is that it will be a lot, and to have a CI duration that is < 1min.

**Comment by quaquel** [comment]
I agree with @Corvince. Let's give UV a few weeks to sort out the initial problems before we consider shifting.

Also, what is the problem with a CI that takes a bit over a minute?

**Comment by rht** [comment]
> Let's give UV a few weeks to sort out the initial problems before we consider shifting.

That sounds like a good timeline for Ruff developers. Ruff used to have a daily (if not more frequent) release cycle. And during this high frequency releases, it already had adopters by huge projects. Faster CI does matter, to the point they have been willing to adopt an alpha-release linter.

> Also, what is the problem with a CI that takes a bit over a minute?

https://github.com/projectmesa/mesa/pull/2039#issuecomment-1948115515

**Comment by EwoutH** [comment]
> but I strongly discourage its use for mesa just yet. The first public release is less than 24 hours ago.

Agreed. Let's wait a few week until uv stabilizes a bit more.

File: .github/workflows/build_lint.yml [hunk 1]
```diff
@@ -60,18 +60,23 @@ jobs:
...
+      run: pip install uv
     - name: Install dependencies
       # Only if the cache misses
       # Based on https://github.com/pypa/pip/issues/8049#issuecomment-633845028
       # read_requirements.py should be removed once
       # https://github.com/pypa/pip/issues/11440 is resolved.
       if: steps.cache.outputs.cache-hit != 'true'
       run: |
-        pip install toml
+        rm -fr ${Python_ROOT_DIR}/lib/python${{ matrix.python-version}}/site-packages/pip-23.0.1.dist-info
```
Comment (Corvince) [inline .github/workflows/build_lint.yml:74]: Why is this line needed? Seems like it would break easily when the pip version changes.

File: .github/workflows/build_lint.yml [hunk 2]
```diff
@@ -60,18 +60,24 @@ jobs:
...
+        rm -fr ${Python_ROOT_DIR}/lib/python${{ matrix.python-version}}/site-packages/pip-23.3.1.dist-info
+        uv pip install toml
         python tests/read_requirements.py > requirements.txt
-        pip install -r requirements.txt
+        uv pip install -r requirements.txt
     - name: Install Mesa
-      run: pip install --no-deps .
+      run: uv pip install --no-deps "mesa @ ."
```
Comment (Corvince) [inline .github/workflows/build_lint.yml:80]: Is the "mesa @" part required? Locally it works just fine without that

**Comment by Corvince** [comment]
I think its funny that pip takes 2 seconds to install just uv and then uv continues to install the dependencies in a few milliseconds. Curious how it will perform with outdated caches.

File: .github/workflows/build_lint.yml (hunk 1 above)
Comment (rht) [reply .github/workflows/build_lint.yml:76]: `uv` breaks if there are multiple `pip` installed. There isn't a bug report for this yet, though if I were to open one, would be resolved within days because of the use case in GH Actions.

File: .github/workflows/build_lint.yml (hunk 2 above)
Comment (rht) [reply .github/workflows/build_lint.yml:85]: Used to require such. It could have been fixed by 0.1.5. Let me try it.

**Comment by rht** [comment]
Everything builds except for Windows, which is a matter of using `rm -r -fo` instead of `rm -fr`.
The macOS build is sped up 2x (which should be sped up further after catching). The examples build is sped up from 27s in #2044 to 17s.

I should be able to revert my advanced (but complex) caching back to `setup-python` pip cache, and it would be simpler and even faster overall.

File: .github/workflows/build_lint.yml (hunk 1 above)
Comment (Corvince) [reply .github/workflows/build_lint.yml]: Are you sure about this one? Specifically I don't see how why multiple pips should be installed at all and why uv should interact with them at all. Also I haven't seen this in other workflow files that already use `uv`

Comment (rht) [reply .github/workflows/build_lint.yml]: That's because they use `uv venv`, e.g. https://github.com/astral-sh/uv/issues/1386#issuecomment-1952420541.

Comment (rht) [reply .github/workflows/build_lint.yml]: I was thinking of waiting until the non-venv solution has stabilized, but `uv venv` is blazing fast, so it shouldn't matter to create the venv.

File: .github/workflows/build_lint.yml (hunk 2 above)
Comment (rht) [reply .github/workflows/build_lint.yml]: Still requires "mesa @ .".

File: .github/workflows/build_lint.yml (hunk 1 above)
Comment (rht) [reply .github/workflows/build_lint.yml]: `uv venv` failed https://github.com/projectmesa/mesa/actions/runs/7980718647/job/21790983916?pr=2038 because Solara has a hardcoded assumption on the virtualenv information.

**Comment by rht** [comment]
Tests all passed.

**Comment by rht** [comment]
It's now even more concise with `uv pip install --system ...`.

**Comment by EwoutH** [comment]
Nice work, I like the reduced complexity of our CI configuration. I think we can try it.

**Comment by rht** [comment]
There has been ~100 PRs opened that contain `uv pip install`, including from the astral-sh repo itself. How about we wait until at least ~300? https://github.com/search?q=%22uv+pip+install%22&type=pullrequests.

File: .github/workflows/build_lint.yml [hunk 3]
```diff
@@ -46,30 +46,11 @@ jobs:
...
-      run: |
-        pip install toml
-        python tests/read_requirements.py > requirements.txt
-        pip install -r requirements.txt
+        cache: 'pip'
+    - name: Install uv
+      run: pip install uv
     - name: Install Mesa
-      run: pip install --no-deps .
+      run: uv pip install --system -e .[dev]
```
Comment (EwoutH) [inline .github/workflows/build_lint.yml:53]: Is the edible install needed here?

Comment (rht) [reply .github/workflows/build_lint.yml]: There are 2 options as described in https://github.com/astral-sh/uv/issues/1945.

Comment (charliermarsh) [reply .github/workflows/build_lint.yml]: I saw this linked from uv -- just as a heads up, this is supported as of `v0.1.24`. You can `uv pip install .` directly, without using an editable.

Comment (rht) [reply .github/workflows/build_lint.yml]: @charliermarsh thank you for the awesome tool! I have amended the PR with `uv pip install --system .[dev]`.

**Review by EwoutH — APPROVED** [review]
```
Run uv pip install --system .[dev]
Resolved 110 packages in 903ms
Downloaded 108 packages in 1.76s
Installed 110 packages in 267ms
```
This is quite awesome. Thanks!
